browser/ui/tabs/brave_tab_strip_model.cc

```diff
@@ -0,0 +1,146 @@
+    : TabStripModel(delegate, profile) {}
+
+BraveTabStripModel::~BraveTabStripModel() {}
+
+void BraveTabStripModel::SelectNextTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
+          ->GetPrefs()
+          ->GetBoolean(kMRUCyclingEnabled);
+
+  if (isMRUEnabled) {
+    SelectTabMRU(false, detail);
+  } else {
+    SelectRelativeTab(true, detail);
+  }
+}
+
+void BraveTabStripModel::SelectPreviousTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
+          ->GetPrefs()
+          ->GetBoolean(kMRUCyclingEnabled);
+
+  if (isMRUEnabled) {
+    SelectTabMRU(true, detail);
+  } else {
+    SelectRelativeTab(false, detail);
+  }
+}
+
+void BraveTabStripModel::SelectTabMRU(bool backward,
+                                      UserGestureDetails detail) {
+  if (current_mru_cycling_index == -1) {
+    // Start cycling
+
+    // Create a list of tab indexes sorted by time of last activation
+    for (int i = 0; i < count(); ++i) {
+      mru_cycle_list.push_back(i);
+    }
+
+    std::sort(mru_cycle_list.begin(), mru_cycle_list.end(),
+              [this](int a, int b) {
+                return GetWebContentsAt(a)->GetLastActiveTime() >
+                       GetWebContentsAt(b)->GetLastActiveTime();
+              });
+
+    current_mru_cycling_index = 0;
+
+    // Create an event handler eating all keyboard events while tabing
+    ctrl_released_event_handler = std::make_unique<CtrlReleaseHandler>(this);
+
+    // Add the event handler
+    gfx::NativeWindow window =
+        this->GetActiveWebContents()->GetTopLevelNativeWindow();
+#if defined(OS_MACOSX)
+    views::Widget::GetWidgetForNativeWindow(window)
+        ->GetRootView()
+        ->AddPreTargetHandler(ctrl_released_event_handler.get());
+#else
+    window->AddPreTargetHandler(ctrl_released_event_handler.get());
+#endif
+  }
+
+  int tabCount = mru_cycle_list.size();
+
+  if (backward) {
+    current_mru_cycling_index =
+        (current_mru_cycling_index - 1 % tabCount + tabCount) % tabCount;
+  } else {
+    current_mru_cycling_index =
+        (current_mru_cycling_index + 1 % tabCount + tabCount) % tabCount;
+  }
+
+  ActivateTabAt(mru_cycle_list[current_mru_cycling_index], detail);
+}
+
+void BraveTabStripModel::StopMRUCycling() {
+  current_mru_cycling_index = -1;
+  mru_cycle_list.clear();
+
+  if (ctrl_released_event_handler) {
+    // Remove the event handler
+    gfx::NativeWindow window =
+        this->GetActiveWebContents()->GetTopLevelNativeWindow();
+#if defined(OS_MACOSX)
+    views::Widget::GetWidgetForNativeWindow(window)
+        ->GetRootView()
+        ->RemovePreTargetHandler(ctrl_released_event_handler.get());
+#else
+    window->RemovePreTargetHandler(ctrl_released_event_handler.get());
+#endif
+
+    ctrl_released_event_handler.reset();
```

Comment:
I can see below crashes frequently while shutdown. Maybe there is some bug for reg/unreg event handler.
```
[16574:775:0410/123156.987706:FATAL:event_handler.cc(26)] Check failed: !check_targets_ || targets_installed_on_.empty().
0   libbase.dylib                       0x000000010f24e21f base::debug::CollectStackTrace(void**, unsigned long) + 31
1   libbase.dylib                       0x000000010ef00fcb base::debug::StackTrace::StackTrace(unsigned long) + 75
2   libbase.dylib                       0x000000010ef0104d base::debug::StackTrace::StackTrace(unsigned long) + 29
3   libbase.dylib                       0x000000010ef01028 base::debug::StackTrace::StackTrace() + 40
4   libbase.dylib                       0x000000010ef4e3d7 logging::LogMessage::~LogMessage() + 183
5   libbase.dylib                       0x000000010ef4d055 logging::LogMessage::~LogMessage() + 21
6   libevents.dylib                     0x000000017099c03a ui::EventHandler::~EventHandler() + 346
7   libchrome_dll.dylib                 0x000000011706a068 BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 24
8   libchrome_dll.dylib                 0x000000011706a085 BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 21
9   libchrome_dll.dylib                 0x000000011706a0ac BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 28
10  libchrome_dll.dylib                 0x000000011706bacf std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler>::operator()(BraveTabStripModel::CtrlReleaseHandler*) const + 47
11  libchrome_dll.dylib                 0x0000000117069fbd std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::reset(BraveTabStripModel::CtrlReleaseHandler*) + 109
12  libchrome_dll.dylib                 0x000000011706a309 std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::~unique_ptr() + 25
13  libchrome_dll.dylib                 0x0000000117069575 std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::~unique_ptr() + 21
14  libchrome_dll.dylib                 0x0000000117069547 BraveTabStripModel::~BraveTabStripModel() + 71
15  libchrome_dll.dylib                 0x0000000117069595 BraveTabStripModel::~BraveTabStripModel() + 21
16  libchrome_dll.dylib                 0x00000001170695bc BraveTabStripModel::~BraveTabStripModel() + 28
17  libchrome_dll.dylib                 0x000000011cec441f std::__1::default_delete<TabStripModel>::operator()(TabStripModel*) const + 47
18  libchrome_dll.dylib                 0x000000011cec439d std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::reset(TabStripModel*) + 109
19  libchrome_dll.dylib                 0x000000011cec4329 std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::~unique_ptr() + 25
20  libchrome_dll.dylib                 0x000000011ceb5375 std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::~unique_ptr() + 21
21  libchrome_dll.dylib                 0x000000011ceb49a2 Browser::~Browser() + 2002
22  libchrome_dll.dylib                 0x000000011ceb53f5 Browser::~Browser() + 21
23  libchrome_dll.dylib                 0x000000011ceb551c Browser::~Browser() + 28
24  libchrome_dll.dylib                 0x000000011d45552f std::__1::default_delete<Browser>::operator()(Browser*) const + 47
25  libchrome_dll.dylib                 0x000000011d4554ed std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::reset(Browser*) + 109
26  libchrome_dll.dylib                 0x000000011d455479 std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::~unique_ptr() + 25
27  libchrome_dll.dylib                 0x000000011d444495 std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::~unique_ptr() + 21
28  libchrome_dll.dylib                 0x000000011d443e5c BrowserView::~BrowserView() + 1836
29  libchrome_dll.dylib                 0x000000011707b4da BraveBrowserView::~BraveBrowserView() + 42
30  libchrome_dll.dylib                 0x000000011707adbc BraveBrowserView::~BraveBrowserView() + 28
31  libchrome_dll.dylib                 0x000000011707adec BraveBrowserView::~BraveBrowserView() + 28
32  libviews.dylib                      0x0000000183a74907 views::View::~View() + 503
33  libviews.dylib                      0x0000000183acf192 views::NonClientView::~NonClientView() + 306
34  libviews.dylib                      0x0000000183acf1dc views::NonClientView::~NonClientView() + 28
35  libviews.dylib                      0x0000000183acf2ec views::NonClientView::~NonClientView() + 28
36  libviews.dylib                      0x000000018385a7af std::__1::default_delete<views::View>::operator()(views::View*) const + 47
37  libviews.dylib                      0x000000018385a72d std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::reset(views::View*) + 109
38  libviews.dylib                      0x000000018385a6b9 std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::~unique_ptr() + 25
39  libviews.dylib                      0x0000000183857235 std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::~unique_ptr() + 21
40  libviews.dylib                      0x0000000183a76a36 views::View::DoRemoveChildView(views::View*, bool, bool, bool, views::View*) + 1638
41  libviews.dylib                      0x0000000183a76ad0 views::View::RemoveAllChildViews(bool) + 112
42  libviews.dylib                      0x0000000183aac744 views::Widget::DestroyRootView() + 100
43  libviews.dylib                      0x0000000183aac4cd views::Widget::~Widget() + 109
44  libchrome_dll.dylib                 0x000000011d434e45 BrowserFrame::~BrowserFrame() + 213
45  libchrome_dll.dylib                 0x000000011708e4e8 BraveBrowserFrame::~BraveBrowserFrame() + 24
46  libchrome_dll.dylib                 0x000000011708e355 BraveBrowserFrame::~BraveBrowserFrame() + 21
47  libchrome_dll.dylib                 0x000000011708e37c BraveBrowserFrame::~BraveBrowserFrame() + 28
48  libviews.dylib                      0x0000000183b131e1 views::NativeWidgetMac::~NativeWidgetMac() + 129
49  libchrome_dll.dylib                 0x000000011d2eae3f BrowserFrameMac::~BrowserFrameMac() + 159
50  libchrome_dll.dylib                 0x000000011d2eae8c BrowserFrameMac::~BrowserFrameMac() + 28
51  libchrome_dll.dylib                 0x000000011d2eaf1c BrowserFrameMac::~BrowserFrameMac() + 28
52  libviews.dylib                      0x0000000183b1363d views::NativeWidgetMac::WindowDestroyed() + 381
53  libviews.dylib                      0x0000000183ae176c views::NativeWidgetMacNSWindowHost::OnWindowHasClosed() + 268
54  libapp_shim.dylib                   0x000000018b1d9b85 remote_cocoa::NativeWidgetNSWindowBridge::OnWindowWillClose() + 1477
55  libapp_shim.dylib                   0x000000018b1f87a0 -[ViewsNSWindowDelegate windowWillClose:] + 720
56  CoreFoundation                      0x00007fff38c6d35f __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
57  CoreFoundation                      0x00007fff38c6d2f3 ___CFXRegistrationPost1_block_invoke + 63
58  CoreFoundation                      0x00007fff38c6d268 _CFXRegistrationPost1 + 372
59  CoreFoundation                      0x00007fff38c6cebe ___CFXNotificationPost_block_invoke + 97
60  CoreFoundation                      0x00007fff38c3c7e2 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1575
61  CoreFoundation                      0x00007fff38c3bc82 _CFXNotificationPost + 1351
62  Foundation                          0x00007fff3b2c1a22 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
63  AppKit                              0x00007fff366d4ec5 -[NSWindow _finishClosingWindow] + 185
64  AppKit                              0x00007fff361b20cd -[NSWindow _close] + 352
65  libapp_shim.dylib                   0x000000018b1d7ce7 ___ZN12remote_cocoa26NativeWidgetNSWindowBridge11CloseWindowEv_block_invoke.203 + 39
66  libapp_shim.dylib                   0x000000018b1e7a78 void base::internal::FunctorTraits<base::mac::ScopedBlock<void () block_pointer>, void>::Invoke<base::mac::ScopedBlock<void () block_pointer> >(base::mac::ScopedBlock<void () block_pointer>&&) + 88
67  libapp_shim.dylib                   0x000000018b1e7a0d void base::internal::InvokeHelper<false, void>::MakeItSo<base::mac::ScopedBlock<void () block_pointer> >(base::mac::ScopedBlock<void () block_pointer>&&) + 29
68  libapp_shim.dylib                   0x000000018b1e79c1 void base::internal::Invoker<base::internal::BindState<base::mac::ScopedBlock<void () block_pointer> >, void ()>::RunImpl<base::mac::ScopedBlock<void () block_pointer>, std::__1::tuple<> >(base::mac::ScopedBlock<void () block_pointer>&&, std::__1::tuple<>&&, std::__1::integer_sequence<unsigned long>) + 33
69  libapp_shim.dylib                   0x000000018b1e7997 base::internal::Invoker<base::internal::BindState<base::mac::ScopedBlock<void () block_pointer> >, void ()>::RunOnce(base::internal::BindStateBase*) + 87
70  libbase.dylib                       0x000000010eeaf385 base::OnceCallback<void ()>::Run() && + 117
71  libbase.dylib                       0x000000010f0ab59b base::TaskAnnotator::RunTask(char const*, base::PendingTask*) + 1563
72  libbase.dylib                       0x000000010f0fcfab base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::sequence_manager::LazyNow*, bool*) + 1627
73  libbase.dylib                       0x000000010f0fc6cf base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoSomeWork() + 159
74  libbase.dylib                       0x000000010f0fd32a non-virtual thunk to base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoSomeWork() + 42
75  libbase.dylib                       0x000000010f2a7600 base::MessagePumpCFRunLoopBase::RunWork() + 160
76  libbase.dylib                       0x000000010f2a773c ___ZN4base24MessagePumpCFRunLoopBase13RunWorkSourceEPv_block_invoke + 28
77  libbase.dylib                       0x000000010f28994a base::mac::CallWithEHFrame(void () block_pointer) + 10
78  libbase.dylib                       0x000000010f2a6583 base::MessagePumpCFRunLoopBase::RunWorkSource(void*) + 115
79  CoreFoundation                      0x00007fff38c77b21 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
80  CoreFoundation                      0x00007fff38c77ac0 __CFRunLoopDoSource0 + 103
81  CoreFoundation                      0x00007fff38c778d4 __CFRunLoopDoSources0 + 209
82  CoreFoundation                      0x00007fff38c76740 __CFRunLoopRun + 1272
83  CoreFoundation                      0x00007fff38c75bd3 CFRunLoopRunSpecific + 499
84  HIToolbox                           0x00007fff377cb65d RunCurrentEventLoopInMode + 292
85  HIToolbox                           0x00007fff377cb39d ReceiveNextEventCommon + 600
86  HIToolbox                           0x00007fff377cb127 _BlockUntilNextEventMatchingListInModeWithFilter + 64
87  AppKit                              0x00007fff35e3bba4 _DPSNextEvent + 990
88  AppKit                              0x00007fff35e3a380 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1352
89  libchrome_dll.dylib                 0x0000000118a31e8a __71-[BrowserCrApplication nextEventMatchingMask:untilDate:inMode:dequeue:]_block_invoke + 106
90  libbase.dylib                       0x000000010f28994a base::mac::CallWithEHFrame(void () block_pointer) + 10
91  libchrome_dll.dylib                 0x0000000118a31d26 -[BrowserCrApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 262
92  AppKit                              0x00007fff35e2c09e -[NSApplication run] + 658
93  libbase.dylib                       0x000000010f2a8bd8 base::MessagePumpNSApplication::DoRun(base::MessagePump::Delegate*) + 312
94  libbase.dylib                       0x000000010f2a5c15 base::MessagePumpCFRunLoopBase::Run(base::MessagePump::Delegate*) + 197
95  libbase.dylib                       0x000000010f0fe1ce base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool, base::TimeDelta) + 830
96  libbase.dylib                       0x000000010f026c8a base::RunLoop::Run() + 858
97  libchrome_dll.dylib                 0x0000000118a5b7a2 ChromeBrowserMainParts::MainMessageLoopRun(int*) + 306
98  libcontent.dylib                    0x00000001435affd9 content::BrowserMainLoop::RunMainMessageLoopParts() + 393
99  libcontent.dylib                    0x00000001435beb0a content::BrowserMainRunnerImpl::Run() + 346
100 libcontent.dylib                    0x00000001435a9b23 content::BrowserMain(content::MainFunctionParams const&) + 211
101 libcontent.dylib                    0x0000000145d61bc2 content::RunBrowserProcessMain(content::MainFunctionParams const&, content::ContentMainDelegate*) + 130
102 libcontent.dylib                    0x0000000145d635bc content::ContentMainRunnerImpl::RunServiceManager(content::MainFunctionParams&, bool) + 1532
103 libcontent.dylib                    0x0000000145d62ecb content::ContentMainRunnerImpl::Run(bool) + 731
104 libcontent.dylib                    0x0000000145d5dee8 content::ContentServiceManagerMainDelegate::RunEmbedderProcess() + 72
105 libembedder.dylib                   0x000000010ea2249f service_manager::Main(service_manager::MainParams const&) + 1903
106 libcontent.dylib                    0x0000000145d61af9 content::ContentMain(content::ContentMainParams const&) + 153
107 libchrome_dll.dylib                 0x0000000114dc5034 ChromeMain + 404
108 Brave Browser Development           0x000000010a136387 main + 583
109 libdyld.dylib                       0x00007fff7031b7fd start + 1
110 ???                                 0x0000000000000004 0x0 + 4
Task trace:
0   libapp_shim.dylib                   0x000000018b1d7833 remote_cocoa::NativeWidgetNSWindowBridge::CloseWindow() + 659
1   libchrome_dll.dylib                 0x000000011ceb904f Browser::TabStripEmpty() + 79

[0410/123157.652279:WARNING:system_snapshot_mac.cc(42)] sysctlbyname kern.nx: No such file or directory (2)
```

Reply:
I could reproduce this by closing the browser while cycling. This is because the `CtrlReleaseHandler` is still registered when shutting down. I solved it by calling a function that deactivate the event handler destructor check. I tried doing a standard unreg but the widget is not accessible anymore when destructing.